### PR TITLE
Proposal for: Flaky port allocation tests #2281

### PR DESCRIPTION
**Pull request by dkhozyainov**
Greetings, please look at my PR for #2281 

## Submitter checklist

- [x] Recommended: Join [WireMock Slack](https://slack.wiremock.org/) to get any help in `#help-contributing` or a project-specific channel like `#wiremock-java`
- [x] Recommended: If you participate in Hacktoberfest 2023, maake sure you're [signed up](https://wiremock.org/events/hacktoberfest/) there and in the WireMock form 
- [x] The PR request is well described and justified, including the body and the references
- [x] The PR title represents the desired changelog entry
- [x] The repository's code style is followed (see the contributing guide)
- [x] Test coverage that demonstrates that the change works as expected
- [x] For new features, there's necessary documentation in this pull request or in a subsequent PR to [wiremock.org](https://github.com/wiremock/wiremock.org)

### CODE REVIEW

**Comment by dkhozyainov** [comment]
@oleg-nenashev @tomakehurst
I tried to take into account this comment:
https://github.com/wiremock/wiremock/pull/2288#issuecomment-1663795711
Please look PR how it will be possible

**Comment by tomakehurst** [comment]
I think there's a much simpler solution to this problem - catch the bind exception specifically in `JettyHttpServer.start()` and retry starting the server (a limited number of times before giving up).

This should also work with parallelised tests, whereas storing port numbers in a local map doesn't help with this as a port number saved as "used" in one process won't be in the others.

**Comment by dkhozyainov** [comment]
ok, I corrected PR, please look it @tomakehurst

File: src/test/java/com/github/tomakehurst/wiremock/jetty11/JettyHttpServerTest.java [hunk 1]
```diff
@@ -119,4 +124,17 @@ public void testHttpConnectorIsNullWhenHttpDisabled()
 
     assertNull(httpConnector);
   }
+
+  @Test
+  public void testStartWithBindException() {
```
Comment (tomakehurst) [inline src/test/java/com/github/tomakehurst/wiremock/jetty11/JettyHttpServerTest.java:129]: I don't think this test case really proves anything.

An alternative would be to:
1. Add an info log statement indicating that a retry is happening and the count value.
2. Create a server socket on port 0
3. Try to start WireMock with this socket's port number and with verbose logging enabled
4. Assert that 3 retry log messages were written
5. Assert that the bind exception is thrown

File: src/test/java/com/github/tomakehurst/wiremock/jetty11/JettyHttpServerTest.java [hunk 2]
```diff
@@ -119,4 +123,26 @@ public void testHttpConnectorIsNullWhenHttpDisabled()
...
+
+    ServerConnector serverConnector = (ServerConnector) testServer.getConnectors()[0];
+    int currentPort = serverConnector.getLocalPort();
+
+    WireMockConfiguration config = WireMockConfiguration.wireMockConfig().port(currentPort);
+    JettyHttpServer jettyHttpServer =
+            (JettyHttpServer)
+                    serverFactory.buildHttpServer(config, adminRequestHandler, stubRequestHandler);
+
+    try {
```
Comment (tomakehurst) [inline src/test/java/com/github/tomakehurst/wiremock/jetty11/JettyHttpServerTest.java:140]: assertThrows might be neater here

**Comment by tomakehurst** [comment]
Apart from the minor test comment, looks good

File: src/test/java/com/github/tomakehurst/wiremock/jetty11/JettyHttpServerTest.java (hunk 2 above)
Comment (dkhozyainov) [reply src/test/java/com/github/tomakehurst/wiremock/jetty11/JettyHttpServerTest.java]: ok, done

File: src/main/java/com/github/tomakehurst/wiremock/jetty/JettyHttpServer.java [hunk 3]
```diff
@@ -192,10 +194,21 @@ protected Handler[] extensionHandlers() {
...
+    int retryCount = 0;
+
+    while (retryCount < MAX_RETRIES) {
+      try {
+        jettyServer.start();
+        break;
+      } catch (IOException bindException) {
+        retryCount++;
+        if (retryCount >= MAX_RETRIES) {
+          throw new RuntimeException(
```
Comment (tomakehurst) [inline src/main/java/com/github/tomakehurst/wiremock/jetty/JettyHttpServer.java:206]: Our coding rules build plugin is (rightly) complaining about throwing a generic exception here. Please can we change this to `FatalStartupException`.

Comment (dkhozyainov) [reply src/main/java/com/github/tomakehurst/wiremock/jetty/JettyHttpServer.java]: ok, done

**Pull request merged**